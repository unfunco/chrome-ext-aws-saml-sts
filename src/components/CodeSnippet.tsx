import * as React from 'react'
import { useState } from 'react'

type CodeSnippetProps = {
  code: string
  ready: boolean
}

const CodeSnippet = ({ code, ready }: CodeSnippetProps): React.ReactElement => {
  const [copied, setCopied] = useState(false)
  const [hovering, setHovering] = useState(false)

  const copyToClipboard = (): void => {
    if ('clipboard' in navigator) {
      navigator.clipboard.writeText(code).then((): void => {
        setHovering(false)
        setCopied(true)
        setTimeout(() => setCopied(false), 2000)
      })
    }
  }

  return (
    <div
      className={`group relative mb-2 border ${
        ready && hovering ? 'cursor-pointer' : ''
      }`}>
      <pre
        className={`overflow-x-scroll border bg-gray-100 p-2`}
        onClick={copyToClipboard}
        onMouseEnter={() => setHovering(true)}
        onMouseLeave={() => setHovering(false)}>
        {ready && <code className={`font-mono`}>{code}</code>}
        {!ready && (
          <span className={'font-sans text-gray-500'}>
            Waiting for credentials...
          </span>
        )}
      </pre>
      {ready && hovering && (
        <div
          className={`left pointer-events-none absolute top-0 flex h-full w-full items-center justify-center bg-black text-base font-bold text-white opacity-75`}>
          Click to copy
        </div>
      )}
      {ready && copied && (
        <div
          className={`pointer-events-none absolute left-1/2 top-0 flex h-full w-full -translate-x-1/2 transform items-center justify-center bg-green-600 text-base font-bold text-white opacity-90`}>
          Copied to clipboard!
        </div>
      )}
    </div>
  )
}

export default CodeSnippet
