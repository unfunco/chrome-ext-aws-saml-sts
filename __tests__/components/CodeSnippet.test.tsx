/** @jest-environment jsdom */

import { act, create, ReactTestRendererJSON } from 'react-test-renderer'
import CodeSnippet from '@/components/CodeSnippet'
import * as React from 'react'

describe('CodeSnippet', (): void => {
  const code = `const greeting = 'Hello, world!'`

  beforeEach((): void => {
    jest.useFakeTimers()
  })

  afterEach((): void => {
    jest.clearAllTimers()
  })

  describe('copy guidance', (): void => {
    it('is not shown by default', (): void => {
      const component = create(<CodeSnippet code={code} />)
      const tree = component.toJSON() as ReactTestRendererJSON
      expect(tree).toMatchSnapshot()
    })

    it('is shown on hover', (): void => {
      const component = create(<CodeSnippet code={code} />)
      let tree = component.toJSON() as ReactTestRendererJSON
      expect(tree).toMatchSnapshot()

      act((): void => {
        const pre = tree.children![0] as ReactTestRendererJSON
        pre.props.onMouseEnter()
      })

      tree = component.toJSON() as ReactTestRendererJSON
      expect(tree).toMatchSnapshot()
    })
  })

  describe('copy confirmation', (): void => {
    it('is not shown by default', (): void => {
      const component = create(<CodeSnippet code={code} />)
      const tree = component.toJSON() as ReactTestRendererJSON
      expect(tree).toMatchSnapshot()
    })
  })
})
