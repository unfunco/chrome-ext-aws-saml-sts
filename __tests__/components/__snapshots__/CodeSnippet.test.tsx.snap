// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeSnippet copy confirmation is not shown by default 1`] = `
<div
  className="group relative mb-2 "
>
  <pre
    className="relative overflow-x-scroll rounded-md border border-gray-300 bg-gray-200 p-2 leading-5 dark:border-gray-700 dark:bg-gray-800"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <code
      className="font-mono"
    >
      const greeting = 'Hello, world!'
    </code>
  </pre>
</div>
`;

exports[`CodeSnippet copy guidance is not shown by default 1`] = `
<div
  className="group relative mb-2 "
>
  <pre
    className="relative overflow-x-scroll rounded-md border border-gray-300 bg-gray-200 p-2 leading-5 dark:border-gray-700 dark:bg-gray-800"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <code
      className="font-mono"
    >
      const greeting = 'Hello, world!'
    </code>
  </pre>
</div>
`;

exports[`CodeSnippet copy guidance is shown on hover 1`] = `
<div
  className="group relative mb-2 "
>
  <pre
    className="relative overflow-x-scroll rounded-md border border-gray-300 bg-gray-200 p-2 leading-5 dark:border-gray-700 dark:bg-gray-800"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <code
      className="font-mono"
    >
      const greeting = 'Hello, world!'
    </code>
  </pre>
</div>
`;

exports[`CodeSnippet copy guidance is shown on hover 2`] = `
<div
  className="group relative mb-2 cursor-pointer"
>
  <pre
    className="relative overflow-x-scroll rounded-md border border-gray-300 bg-gray-200 p-2 leading-5 dark:border-gray-700 dark:bg-gray-800"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <code
      className="font-mono"
    >
      const greeting = 'Hello, world!'
    </code>
  </pre>
  <div
    className="left pointer-events-none absolute top-0 flex h-full w-full items-center justify-center rounded-md bg-black text-base font-bold text-white opacity-75 dark:bg-gray-100 dark:text-black"
  >
    Click to copy
  </div>
</div>
`;
